design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+ */
+public interface Row {
+
+  List<String> columnNames();
+
+  List<String> columnTypes();
+
+  List<Object> values();
+
+  /**
+   * Get the value for a particular column of the Row as an Object.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Object getObject(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as an Object.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Object getObject(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a string.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  String getString(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a string.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  String getString(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as an integer.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Integer getInt(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as an integer.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Integer getInt(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Long getLong(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Long getLong(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Double getDouble(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Double getDouble(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Boolean getBoolean(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Boolean getBoolean(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  BigDecimal getDecimal(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  BigDecimal getDecimal(String columnName);
+}
```

Comment:
Should there be helpers for nested types, too? Or is `getObject()` plus casting the intended way to handle them?

Reply:
For struct types, if we return a JsonObject that is a nested data structure.

Reply:
But there is no method with that return type. Thus, I am asking if we should add a corresponding method, to avoid casting from `Object` to `JsonObject` (even if I am not sure, if we should return `JsonObject` but a kslqDB "native" type -- cf. the other comments).

Reply:
Sure, I believe the intention is to do that.

Reply:
\cc @vcrfxia ?

Reply:
Maybe getStruct() would be better than getKsqlObject() ?

Reply:
`getStruct()` seems it could be confusing since the same `KsqlObject` type is used to represent both the STRUCT and MAP ksql types. 

(To clarify: I'm not a fan of the name `getKsqlObject()` -- or `KsqlObject` for that matter -- either, but I'm not sure `getStruct()` is an improvement.)

Reply:
getJsonObject() ?

Reply:
Let's discuss in https://github.com/confluentinc/ksql/pull/5327 where I've added the new types. I haven't exposed the JSON methods on `KsqlObject` and `KsqlArray` in order to keep the exposed methods to a minimum, but it sounds like you prefer to expose those methods and rename the getters to `getJsonObject()` and `getJsonArray()`, which I'm happy with if we think the JSON methods are worth exposing.